File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/collision/detector/box/BoxVehicleCollisionsDetector.java
package com.github.aadvorak.artilleryonline.battle.collision.detector.box;

import com.github.aadvorak.artilleryonline.battle.calculations.BattleCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.BoxCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.Calculations;
import com.github.aadvorak.artilleryonline.battle.collision.Collision;
import com.github.aadvorak.artilleryonline.battle.collision.detector.CollisionsDetector;
import com.github.aadvorak.artilleryonline.battle.common.lines.BodyPart;
import com.github.aadvorak.artilleryonline.battle.common.lines.Circle;
import com.github.aadvorak.artilleryonline.battle.utils.ContactUtils;
import org.springframework.stereotype.Component;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

@Component
public class BoxVehicleCollisionsDetector implements CollisionsDetector {

    @Override
    public Set<Collision> detect(Calculations<?> calculations, BattleCalculations battle, boolean first) {
        if (calculations instanceof BoxCalculations boxCalculations) {
            return detect(boxCalculations, battle, first);
        }
        return Set.of();
    }

    private Set<Collision> detect(BoxCalculations box, BattleCalculations battle, boolean first) {
        Set<Collision> collisions = new HashSet<>();
        var maxRadius = box.getModel().getPreCalc().getMaxRadius();
        var bodyPart = BodyPart.of(box.getGeometryNextPosition(), box.getModel().getSpecs().getShape());
        for (var vehicle : battle.getVehicles()) {
            var otherMaxRadius = vehicle.getModel().getPreCalc().getMaxRadius();
            var otherPosition = vehicle.getGeometryNextPosition();
            if (otherPosition.getCenter().distanceTo(box.getNext().getPosition().getCenter())
                    > maxRadius + otherMaxRadius) {
                continue;
            }
            var otherWheelRadius = vehicle.getModel().getSpecs().getWheelRadius();
            var otherLeftWheelPosition = vehicle.getLeftWheel().getNext().getPosition();
            var otherRightWheelPosition = vehicle.getRightWheel().getNext().getPosition();
            var otherParts = Map.of(
                    BodyPart.of(otherPosition, vehicle.getModel().getSpecs().getTurretShape()), vehicle,
                    new Circle(otherLeftWheelPosition, otherWheelRadius), vehicle.getLeftWheel(),
                    new Circle(otherRightWheelPosition, otherWheelRadius), vehicle.getRightWheel()
            );
            for (var otherPart : otherParts.entrySet()) {
                var contact = ContactUtils.getBodyPartsContact(bodyPart, otherPart.getKey());
                if (contact != null) {
                    collisions.add(Collision.withVehicle(box, otherPart.getValue(), contact));
                    if (first) return collisions;
                }
            }
        }
        return collisions;
    }
}

File: frontend/playground/battle/collision/detector/box-vehicle-collisions-detector.ts
import {Collision, type CollisionsDetector} from "~/playground/battle/collision/collision"
import {
  BattleCalculations, type BodyCalculations,
  BoxCalculations,
  type Calculations
} from "~/playground/data/calculations"
import {type BodyPart, Circle} from "~/playground/data/geometry"
import {ContactUtils} from "~/playground/utils/contact-utils"
import {BattleUtils} from "~/playground/utils/battle-utils"
import {BodyUtils} from "~/playground/utils/body-utils"

export class BoxVehicleCollisionsDetector implements CollisionsDetector {
  detect(calculations: Calculations, battle: BattleCalculations): Set<Collision> {
    if (calculations instanceof BoxCalculations) {
      return this.detectForBox(calculations as BoxCalculations, battle)
    }
    return new Set()
  }

  private detectForBox(box: BoxCalculations, battle: BattleCalculations): Set<Collision> {
    const collisions = new Set<Collision>()
    const maxRadius = box.model.preCalc.maxRadius
    const bodyPart = BodyUtils.getBodyPart(box.model.specs.shape, box.getGeometryNextPosition()!)
    
    for (const vehicle of battle.vehicles) {
      const otherMaxRadius = vehicle.model.preCalc.maxRadius
      const otherPosition = vehicle.getGeometryNextPosition()!
      
      if (BattleUtils.distance(otherPosition, box.getNextPosition()!) > maxRadius + otherMaxRadius) {
        continue
      }
      
      const otherWheelRadius = vehicle.model.specs.wheelRadius
      const otherLeftWheelPosition = vehicle.leftWheel.nextPosition!
      const otherRightWheelPosition = vehicle.rightWheel.nextPosition!
      
      const otherParts = new Map<BodyPart, BodyCalculations>([
        [BodyUtils.getBodyPart(vehicle.model.specs.turretShape, otherPosition), vehicle],
        [new Circle(otherLeftWheelPosition, otherWheelRadius), vehicle.leftWheel],
        [new Circle(otherRightWheelPosition, otherWheelRadius), vehicle.rightWheel]
      ])
      
      for (const [otherPart, otherPartValue] of otherParts) {
        const contact = ContactUtils.getBodyPartsContact(bodyPart, otherPart)
        if (contact !== null) {
          collisions.add(Collision.withVehicle(box, otherPartValue, contact))
        }
      }
    }
    
    return collisions
  }
}
